### DISCUSSION TITLE
Adding New Folders to C Projects

### DISCUSSION THREAD

**Discussion by gregwillits** [eclipse-cdt/cdt#221] [Q&A] [answered]
Currently using 2022-09 (4.25.0) / CDT 10.7.1 / macOS 10.15.7

Seeking clarification:

This has confused me forever. Neither docs, experiments, nor Google have been able to shed light.

Whenever I add a new folder within the source code project, I have to go into project Properties > C/C++ Build > Tool Settings > GCC C Compiiler > Includes and manually add the full folder path into the list.

I have tried use the project contextual menu to add "Source Folder," but that doesn't even add the folder where I want it. Regardless of the level of parent folder I select to add a folder to, the Source Folder option always puts the folder at the root of the project. So, obviously that has some special purpose.

It just seems to me that adding a folder, just like adding a new header file, should be something that is added to the project automatically.

Is this manual intervention for folders really required, or have I somehow missed configuring something correctly?

Thanks.
-- greg

**Comment by jonahgraham** [comment, marked as answer]
I think there may be some terminology confusion here, not quite sure.

A "source folder" in CDT is a folder that the sources are in, these are folders that are indexed, including for build, etc.

They can be configured in the [project properties]( https://help.eclipse.org/latest/topic/org.eclipse.cdt.doc.user/reference/cdt_u_prop_general_pns_src.htm?cp=9_4_7_0_4_5_4)

These source folders appear at the top level all the time.

Includes need to be somewhat manually managed if you want that folder to be explicitly listed as a `-I` folder. This is done where you are describing, in the "Includes" tab under [C/C++ Build](https://help.eclipse.org/latest/topic/org.eclipse.cdt.doc.user/reference/cdt_u_prop_build_settings_tool.htm?cp=9_4_7_0_3_3_0)

What I think you want to enter in the Includes tab is something like `"${workspace_loc:/${ProjName}/src/inc}"`. The magic bits should be filled in automatically by pressing the Workspace button when adding a new include.

Here is a screenshot showing all of it in one place.

![image](https://user-images.githubusercontent.com/679236/207167466-e1d7a3b2-f60c-43b9-b969-d17d1c0b6ced.png)

Hopefully the above either answers your question, or provides the starting point to ask me a clarifying question.

**Reply by gregwillits** [reply]
Re: include folders -- yep, that's what I've been doing. I just wanted to verify that it did indeed have to be manually done.

I was thinking there had to be a way for the IDE to scan my top level folder and assume that everything in there was part of the build unless I excluded it.

I never have used a Source Folder, so still not sure how it is different or what capabilities it adds. Docs don't explain (that I see) what one is — just how to edit its configuration. I've always just created a normal folder, added it as the first entry in includes, and away we go.

Thanks for the details.

**Reply by jonahgraham** [reply]
There is some (dated) help on source folders [here](https://help.eclipse.org/latest/topic/org.eclipse.cdt.doc.user/tasks/cdt_t_set_src_fold.htm)

> I've always just created a normal folder, added it as the first entry in includes, and away we go.

That is what you are supposed to do. Sorry it isn't more obvious that that wasn't obvious enough, though glad to have helped.

**Comment by gregwillits** [comment]
OH — one more related thing to the Includes paths — whenever I copy a whole project (all the hidden project settings files included) to another computer (same OS, same version of Eclipse CDT) whether manually or from a version control system, I have to manually re-enter every single folder in the whole project into the Includes list (quite a chore for big projects). Given that I am using workspace relative paths, and essentially everything else in the project seems to move, I am expecting that the Includes list would too.

Is that the expected behavior, or is something not set correctly?

Thx.

**Reply by jonahgraham** [reply]
Those include paths are supposed to be in the `.cproject` file. You definitely should not have to recreate them. If you can provide some reproducible steps please file an issue.

**Reply by gregwillits** [reply]
I updated both machines to the latest 2022-12/11.0.0.
Then I started a detailed writeup in Issues with a new project, screenshots, etc.
Moved it to the new machine. Of course, it worked. What. Ever.  :-) 
I copied a major existing project and that worked too.
OK, fine. Either something changed, or I did something differently that I haven't done in 10 other times trying.

Anyway, my only niggle now is that after importing a major project into the new machine, the Project Explorer is labeling the project differently. Editing the project name doesn't change it. Screencap attached -- the brown `(in rcmMain)` is not part of the name, nor does it show up on the originating maching project explorer. How do I get rid of that?

Thanks for your time Jonah.

<img width="213" alt="ProjectName" src="https://user-images.githubusercontent.com/248531/208266143-44bd3d01-f35c-4ee8-b3da-f4ae66a5c6af.png">

**Reply by jonahgraham** [reply]
The brown bit `(in rcmMain)` is a label decoration. Normally that looks like source control to me, on my machine it will show the repo name + branch name in square brackets:

![image](https://user-images.githubusercontent.com/679236/208268009-50502581-d57d-4888-af49-0b124c5220a9.png)

But I don't recognize the specific format of `(in ...)` you have.

I don't know if there is a cleverer way to figure this out, but you can toggle each label decoration in *Preferences* -> *General* -> *Appearance* -> *Label Decorations*:

![image](https://user-images.githubusercontent.com/679236/208268042-83ddbd93-23b5-4165-b802-c6c963ba19b7.png)

If it is the git one specifically that is contributing it, you can customize it in *Preferences* -> *Version Control (Team)* -> *Git* -> *Label Decorations* -> *Text Decorations* tab:

![image](https://user-images.githubusercontent.com/679236/208268077-0c6924cc-4929-4622-80c7-e250b66b768c.png)

**Reply by jonahgraham** [reply]
> Moved it to the new machine. Of course, it worked. What. Ever. :-)

I am glad it worked this time, sad that you lost some time and that we don't have a use case.

**Reply by gregwillits** [reply]
Label Decorations — the two systems are set up the same way. On the imported project, I turned every decoration off in General and Git (which this project is using SVN BTW). It still shows up. I wonder if it is related to the importing.

The path is ....EclipseWS/rcmCoreG3/rcmMain (there's other folders at the same level of rcmMain). Even though the import named the project rcmCoreG3, something wants to identify that the real project level is rcmMain (which is where the .cproject file is). But it's structured this way on both machines.

I guess I'll experiment with imports. Sorry, didn't expect this turn into a thing.

**Reply by jonahgraham** [reply]
That is weird. Project names in the workspace has slightly different rules then projects which on disk are outside of the workspace. For example eclipse tries to keep name of project matching name of folder when project is in the workspace folder, but not so much outside.

Anyway you have asked questions  outside of  my knowledge. I look forward to learning with you as you discover what is going on.
